### DISCUSSION TITLE
Is it possible to fetch a workflow variable through API?

### DISCUSSION THREAD

**Discussion by ccyen8358** [elsa-workflows/elsa-core#822] [Q&A]
Hi,

I just read the [document approval](https://elsa-workflows.github.io/elsa-core/docs/guides-document-approval#sendemail) sample guide. In this guide we use the `signalUrl` function to get the url that triggers the `Signaled` Activity from client side (email in this case).

Is it possible to get this url through an api? 
More broadly, is it possible to fetch any workflow variable through an api?

The reason I ask this question is because we want to embedded the "Signaled url" in our front-end SPA app instead of in the email link. 

Sorry for taking up your time and thanks in advance.

**Comment by DavidPx** [comment]
The activity execution context lets you manipulate variables so I think you'd just set the signal URL like so:

```
using Elsa.Activities.Http.Extensions;
...
.Then(x =>
                {
                    var signalUrl = x.GenerateSignalUrl("foo");
                    x.SetVariable("mySignalUrl", signalUrl);
                })
```

And fetching variables from running workflows works great.  Just get your `WorkflowInstance` from an `IWorkflowInstanceStore` and then hit up its `Variables` property.  It has a typed Get method which takes the variable name.

**Reply by sfmskywalker** [reply]
That's a great way to generate a signal URL and reuse it later on. This also demonstrates the `GenerateSignalUrl` API exposed as an extension method on `ActivityExecutionContext`.

However, this is all Elsa 2 - in Elsa 1, the signal URL generation logic is not publically exposed unfortunately. However, there's no reason it couldn't be exposed, and I think we should do so in Elsa 2 for precisely the use case described by the OP.
